src/datachain/lib/signal_schema.py

```diff
@@ -464,19 +464,63 @@ def contains_file(self) -> bool:
         return False
 
     def slice(
-        self, keys: Sequence[str], setup: Optional[dict[str, Callable]] = None
+        self,
+        params: dict[str, Union[DataType, Any]],
+        setup: Optional[dict[str, Callable]] = None,
+        is_batch: bool = False,
     ) -> "SignalSchema":
-        # Make new schema that combines current schema and setup signals
-        setup = setup or {}
-        setup_no_types = dict.fromkeys(setup.keys(), str)
-        union = SignalSchema(self.values | setup_no_types)
-        # Slice combined schema by keys
-        schema = {}
-        for k in keys:
+        """
+        Returns new schema that combines current schema and setup signals.
+        """
+        setup_params = setup.keys() if setup else []
+        schema: dict[str, DataType] = {}
+
+        for param, param_type in params.items():
+            if param in setup_params:
+                schema[param] = str
+                continue
+
             try:
-                schema[k] = union._find_in_tree(k.split("."))
+                schema_type = self._find_in_tree(param.split("."))
             except SignalResolvingError:
-                pass
+                continue
+
+            if param_type is Any:
+                schema[param] = schema_type
+                continue
+
+            schema_origin = get_origin(schema_type)
+            param_origin = get_origin(param_type)
```

Comment:
We will use it couple times below, so just cache the result of `get_origin` here.